File: javaRevision/array/Sort_0_1.java
package array;

public class Sort_0_1 {
    public static void main(String[] args) {
        int[] arr = { 1, 0, 0, 1, 1, 0, 1, 1, 0 };
        sort(arr);
    }

    public static void sort(int[] arr) {
        int left = 0, right = arr.length - 1;
        while (left < right) {
            // check left for 0
            while(arr[left] == 0 && left<right)
                left++;
            // check right for 1
            while(arr[right] == 1 && left<right)
                right--;

            if(arr[left]==1 && arr[right]==0 && left<right){
                arr[left++] = 0;
                arr[right--] = 1;
            }
        }
        for (int i : arr)
            System.out.print(i + " ");
    }
}

File: c++/GEM_QUESTION/sort_0_1.cpp
#include<iostream>
using namespace std;
void sort(int arr[],  int size)
{
    int left=0, right=size-1;
    while(left<right)
    {
        while(arr[left]==0 && left<right)
        {
            left++;
        }
        while(arr[right]==1 && left<right)
        {
            right--;
        }
        while(arr[left]==1 && arr[right]==0 && left<right)
        {
            swap(arr[left], arr[right]);
            left++;
            right--;
        }
    }
}
void print(int arr[], int size)
{
    cout<<"array after sort: "<<endl;
    for(int i=0; i<size ; i++)
    {
        cout<<arr[i]<<" ";
    }
    cout<<endl;
}

int main()
{
    int arr[100], size;
    cout<<"enter size: ";
    cin>>size;
    cout<<"Enter elements: ";
    for(int i=0; i<size; i++)  
        cin>>arr[i];

    sort(arr, size);
    print(arr, size);

    return 0;
}
